webui/react/src/pages/TrialDetails/TrialDetailsOverview.tsx

```diff
@@ -22,9 +27,49 @@ const TrialDetailsOverview: React.FC<Props> = ({ experiment, trial }: Props) =>
     updateSettings,
   } = useSettings<Settings>(settingsConfig, { storagePath });
 
-  const { defaultMetrics, metricNames, metrics } = useMemo(() => {
+  const [ metricNames, setMetricNames ] = useState<MetricName[]>([]);
+
+  // Stream available metrics.
+  useEffect(() => {
+    const canceler = new AbortController();
+    const trainingMetricsMap: Record<string, boolean> = {};
+    const validationMetricsMap: Record<string, boolean> = {};
+
+    readStream<V1MetricNamesResponse>(
+      detApi.StreamingInternal.metricNames(
+        experiment.id,
+        undefined,
+        { signal: canceler.signal },
+      ),
+      (event) => {
+        if (!event) return;
+        /*
+         * The metrics endpoint can intermittently send empty lists,
+         * so we keep track of what we have seen on our end and
+         * only add new metrics we have not seen to the list.
+         */
+        (event.trainingMetrics || []).forEach((metric) => trainingMetricsMap[metric] = true);
+        (event.validationMetrics || []).forEach((metric) => validationMetricsMap[metric] = true);
+        const newTrainingMetrics = Object.keys(trainingMetricsMap).sort(alphaNumericSorter);
+        const newValidationMetrics = Object.keys(validationMetricsMap).sort(alphaNumericSorter);
+        const newMetrics = [
+          ...(newValidationMetrics || []).map((name) => ({ name, type: MetricType.Validation })),
+          ...(newTrainingMetrics || []).map((name) => ({ name, type: MetricType.Training })),
+        ];
+        setMetricNames(newMetrics);
```

Comment:
**nit perf question:** with this current way, everything time the event callback is called, will trigger `metricNames` to update, and if any of the JSX is dependent on it, it will re-render upon each event callback call. One improvement would be to only updating the metrics if there is a change to the existing metric names.

Reply:
replacing with
```
if (metricNames.length !== newMetrics.length) {
          setMetricNames(newMetrics);
}
```